crates/std_detect/src/detect/cache.rs

```diff
@@ -124,61 +133,82 @@ impl Cache {
 #[cfg(target_pointer_width = "32")]
 struct Cache(AtomicU32, AtomicU32);
 
+#[cfg(target_pointer_width = "32")]
+const UNINITIALIZED_VALUE: u32 = u32::max_value();
+
+#[cfg(target_pointer_width = "32")]
+const UNINITIALIZED_BIT: u32 = 31;
+
+/// Maximum number of features that can be cached.
+/// We reserve a single bit in each of the two atomic values
+/// to check if the cache is initialized
+#[cfg(target_pointer_width = "32")]
+const CACHE_CAPACITY: u32 = 62;
+
 #[cfg(target_pointer_width = "32")]
 impl Cache {
     /// Creates an uninitialized cache.
     const fn uninitialized() -> Self {
         Cache(
-            AtomicU32::new(u32::max_value()),
-            AtomicU32::new(u32::max_value()),
+            AtomicU32::new(UNINITIALIZED_VALUE),
+            AtomicU32::new(UNINITIALIZED_VALUE),
         )
     }
-    /// Is the cache uninitialized?
-    #[inline]
-    pub(crate) fn is_uninitialized(&self) -> bool {
-        self.1.load(Ordering::Relaxed) == u32::max_value()
```

Comment:
So IIUC the problem was here right? This checks whether the second atomic has been initialized, and if that is the case, it assumes both atomics to be initialized. However, the function that initializes the two atomics below performs two relaxes writes, and while this test returns true if the second write has happened, there is no guarantee that the first write has happened, and therefore the UB. Is that correct?

Reply:
If so, wouldn't it suffice to have the first write happen before the second write ?

Reply:
Yes. This could also be fixed by using `Acquire` here and `Release` when doing `self.1.store` further down. However doing it this way also allows you to get away with just a single atomic load on the fast path.

Reply:
If you wish to do it with `Release` semantics, then the fast path would contain a single load with `Release` semantics and a single load with `Relaxed` semantics. This PR only uses a single `Relaxed` load on the fast path.

Reply:
We currently have this code:

```rust
// thread 1
self.0.store(rel);
self.1.store(rel); 

// thread 2
self.1.load(rel); // assumes self.0.store(rel) has happened
```

Is there a way to have `self.0.store` happen-before `self.1.store` such that a relaxed load in thread 2 of `self.1.load` can only observe the initialized state if both stores have happened ?

Reply:
`self.0.store` is already happen-before `self.1.store` given that they are run in the same thread. I am not sure I understand your question.

There are two options as far as I can tell:
- Have a single bit that signifies in one of the atomics that initialization is done. You first write all the other data, then write this bit using release-semantics. If you ever observe this bit being set using acquire-semantics, then you can use relaxed-semantics to get all the other values.
- Have a single bit in each atomic value. That way you can load and check them independently of each other and you will only ever need relaxed loads and stores.

Reply:
> Have a single bit that signifies in one of the atomics that initialization is done. You first write all the other data, then write this bit using release-semantics. If you ever observe this bit being set using acquire-semantics, then you can use relaxed-semantics to get all the other values.

So then I don't think I'm understanding the issue. Isn't this what we currently do?

Reply:
Currently we have:

```rust
// compile-time:
AtomicU32 x, y = {0xFFFFFFFF, 0xFFFFFFFF);

// run-time:
if test(y.load(relaxed) == 0xFFFFFFFF) {
      x.store(relaxed) = ..; 
      y.store(relaxed) = ..; // this bit pattern is never == 0xFFFFFFFF }
}
```

Since x.store happens-before the y.store, the only way for a thread to observe an `y.load(relaxed) != 0xFFFFFFFF` is if there has been a write to both `x` and `y`, right?

Reply:
What this allows is something like this to happen: 

```
y.load(relaxed) == 0xFFFFFFFF // thread 1
x.store(relaxed) = ...; // thread 1
y.load(relaxed) == 0xFFFFFFFF // thread 2
x.store(relaxed) = ...; // thread 2
y.store(relaxed) = ...; // thread 2
y.store(relaxed) = ...; // thread 1
```

where two threads both end up writing to x and y. But that's ok AFAICT.

Reply:
Cant the compiler reorder the store of y before the store of x, because both are relaxed?

Reply:
That's why I asked above if there is a way to make the store of x happen before the store of y. That's all we would need here for correctness AFAICT.

Reply:
https://bartoszmilewski.com/2008/12/01/c-atomics-and-memory-ordering/

> memory_order_release: preceding stores are not moved past the current store or any subsequent stores.

Reply:
So would it suffice to change this to:

```rust
// run-time:
if y.load(relaxed) == 0xFFFFFFFF) {
      x.store(relaxed) = ..; 
      y.store(release) = ..never 0xFFFFFFFF..;
}
```

?

Reply:
@bjorn3 that blog post also says:

> the x86 guarantees acquire semantics for loads and release semantics for stores

so... maybe it only matters for compiler re-orderings, but we don't really want any here anyways. So we can probably just write:

```rust
// run-time:
if y.load(acquire) == 0xFFFFFFFF) {
      x.store(relaxed) = ..; 
      y.store(release) = ..never 0xFFFFFFFF..;
}
```

and call it a day.

Reply:
What we have currently is roughly equivalent to this:

https://play.rust-lang.org/?version=stable&mode=release&edition=2018&gist=81bab903b3535435ad9f33afbcfaf252

Your solution with using acquire and release is correct basically what I suggested in my first option. The downside of it is that you need to use acquire and release.

Reply:
@TethysSvensson I think it is more like [this](https://play.rust-lang.org/?version=stable&mode=release&edition=2018&gist=d64b6a6d105924a8633912b9db6aba44) (we never change the value of the atomics after the first assignment). 

I also tried to add Acquire / Release to your example without much luck, I still get a bad count.

Reply:
I cannot get my example to work either. I think it was broken.

You are right that the value is never reset in the current code. I only reset it to run the experiment multiple times.

The point I am trying to make is that there is currently nothing introduces a synchronization point across threads. So even though it works on x86 which has a very strong memory model, I think the code can give the wrong result when only relying on the LLVM memory model.

Reply:
cc @RalfJung

Reply:
> The point I am trying to make is that there is currently nothing introduces a synchronization point across threads

Yes, I am able to follow this much. I also understand that Release+Acquire would suffice here. 

What I'm still not sure about is why do we need the Acquire at all? IIUC having a relaxed store followed by a release store, and then doing a relaxed load on the second store, should be enough to guarantee that both stores happen before the relaxed load that observes the initialized cache.

Reply:
As far as the C11 memory model (and thus presumably the primary target of the llvm memory model) goes, a release without an acquire is meaningless. As a practical example, the problematic code is actually something like 
```rust
if y.load(relaxed) == 0xFFFFFFFF) {
      x.store(relaxed) = foo; 
      y.store(release) = bar; //..never 0xFFFFFFFF..
}
let r = x.load(relaxed);
```

and the `x.load(relaxed)` could be moved before the if statement, so long as it's corrected for the write inside of the branch, ie it could be:

```rust
let mut r = x.load(relaxed);
if y.load(relaxed) == 0xFFFFFFFF) {
      r = foo;
      x.store(relaxed) = r; 
      y.store(release) = bar; //..never 0xFFFFFFFF..
}
```

making it be `y.load(acquire)` prevents this

Reply:
I'd personally always use release/acquire unless it is either blatantly obvious that relaxed is correct (that's clearly not the case here) or we have benchmarks showing that release/acquire is too expensive.